arch/arm64/core/thread.c

```diff
@@ -124,9 +124,25 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 
 void *z_arch_get_next_switch_handle(struct k_thread **old_thread)
 {
+	/*
+	 * When returning from this function we will have the current thread
+	 * onto the stack to be popped in x1 and the next thread in x0 returned
+	 * from z_get_next_switch_handle() (see isr_wrapper.S)
+	 */
 	*old_thread =  _current;
 
-	return z_get_next_switch_handle(*old_thread);
+	/*
+	 * XXX: see thread in #41840 and #40795
+	 *
+	 * We pass NULL here to _not_ set the switch_handle for the old thread
+	 * at the end of the z_get_next_switch_handle() to avoid a race
+	 * condition with other cores waiting for that on wait_for_switch().
+	 * Instead we set the switch_handle _only_ in the
+	 * z_arm64_context_switch() code when we are 100% sure that the context
+	 * has been swapped out and the thread can be safely picked up by a
+	 * different core.
```

Comment:
I think this would be more understandable if you explain this as an optimization on top of the expectations of the original API, something like:


```c
/* The scheduler API requires a complete switch handle here, but arm64
 * optimizes things such that the callee-save registers are still
 * unsaved here (they get written out in z_arm64_context_switch()
 * below).  So pass a NULL instead, which the scheduler will store
 * into the thread switch_handle field.  The resulting thread won't be
 * switched into until we write that ourselves.
 */
```